Add split_by_apid function and command line interface

ddasilva opened the pull request:
This merge request adds a function split_by_apid() which loops through packets in a file of mixed APIDs and sorts the packets into a dictionary where the keys are the integer APID and the values are homogenous streams for each APID. 

There is also a new command line interface exposed through `python -m ccsdspy` which allows splitting a file and writing individual files.
```
   $ python -m ccsdspy split mixed_file.tlm
   Parsing done!
   Writing ./apid00132.tlm...
   Writing ./apid00134.tlm...
   Writing ./apid00258.tlm...
   Writing ./apid00384.tlm...
   Writing ./apid00385.tlm...
   Writing ./apid00386.tlm...
   Writing ./apid00387.tlm...
```

jmbhughes commented on docs/index.rst:
```diff
@@ -48,6 +48,31 @@ The result is returned as an OrderedDict, containing PacketField names as keys a
+
+  import ccsdspy
+
+  with open('mixed_file.tlm', 'rb') as mixed_file)
```
Typo here. The line should end with a colon instead of a parenthesis.

jmbhughes commented on ccsdspy/utils.py:
```diff
@@ -0,0 +1,66 @@
+    
+    Parameters
+    ----------
+    mixed_file: str
```
The type annotation of str isn't complete. On the next line you note in longer form English that it could be a file-like object. It should probably be noted on this line. Otherwise, people using PyCharm like me get a warning that we're passing in the wrong type.

ddasilva replied:
What is the convention for multiple types for PyCharm here? would `mixed_file: str, file-like` work?

jmbhughes replied:
I think you can just type `mixed_file: str or file-like`. Now that I think about it, maybe PyCharm was inferring the type from the code instead of the docstring. I'm not sure. I tend to use [type hints](https://peps.python.org/pep-0484/) everywhere so it doesn't have to do much inference.

jmbhughes commented on docs/index.rst:
```diff
@@ -48,6 +48,31 @@ The result is returned as an OrderedDict, containing PacketField names as keys a
+.. code-block:: python
+
+  import ccsdspy
+
```
Add: `from ccsdspy.util import split_by_apid`, otherwise the code example doesn't run.

jmbhughes commented on ccsdspy/utils.py:
```diff
@@ -0,0 +1,66 @@
+    mixed_file: str
+       Path to file on the local file system, or file-like object
+    valid_apids: list of int, None
+       Optional list of valid APIDs. If specified, warning will be issued when
```
I really like this feature!

jmbhughes commented on ccsdspy/utils.py:
```diff
@@ -0,0 +1,66 @@
+        apid &= 0x07FF
+                
+        if valid_apids is not None and apid not in valid_apids:
+            warnings.warn('Found unknown APID %d' % apid)
```
This is technically fine, but I think an f-string is generally preferred now (although it does require Python 3.6 or higher I believe). So maybe replace with `warnings.warn(f'Found unknown APID {apid}')`. I think there are few spots in the new code that do this. 

Some packages, SunPy for example dictate not to use the old % formatting convention in their [style guides](https://docs.sunpy.org/en/latest/dev_guide/contents/code_standards.html#language-standard).

ddasilva replied:
Technically the last release still supported 2.7!... but it's probably time we can cut those people off. This is a good idea and I'll move towards f-strings

jmbhughes commented on docs/index.rst:
```diff
@@ -48,6 +48,31 @@ The result is returned as an OrderedDict, containing PacketField names as keys a
 
+Splitting Mixed Streams by APID
+-------------------------------
+Often, CCSDS data with arrive from external sources into software systems in a single file with multiple APID's. Splitting a mixed file or stream of bytes by APID so they can be used with the `ccsdspy.FixedLength` class can be done thorugh the API or with the module command line interface.
```
A couple very minor typos. Replace with: 

Often, CCSDS data with arrive from external sources into software systems in a single file with multiple APIDs. Splitting a mixed file or stream of bytes by APID so they can be used with the `ccsdspy.FixedLength` class can be done through the API or with the module command line interface.

jmbhughes reviewed (changes requested):
This is a great addition! It solves the issue I was facing. 

My only major concern is that there aren't any tests for it. I see in your other tests you have some data. We could construct a test by just appending some of the existing data together and then trying to split it. 

Everything else I noted here is pretty much very minor typos.

ehsteve commented:
I also agree that this is a great addition! This seems to require that each of the packets be fixed length since the packets are parsed but simply splitting the stream up and outputting the stream into individual files should not require parsing the packets only finding the edges of the packets. I think this functionality would be more versatile if it did not require the stream to only include fixed length packets.

ehsteve commented:
I agree with @jmbhughes that there should be tests. I manage the [HERMES SOC Org.](https://github.com/HERMES-SOC) and we've set up automatic code coverage (see [here](https://github.com/HERMES-SOC/hermes_core#readme) for an example. I'd be happy to help get it set up if you're interested.

ehsteve commented on ccsdspy/interface.py:
```diff
@@ -99,7 +99,7 @@ def load(self, file):
         Returns
         -------
-        `OrderedDict` mapping field names to NumPy arrays.
+        sorted dictionary mapping field names to NumPy arrays.
```
This states that the dictionary is sorted but I don't see any code that actually sorts the dictionary. For clarity I would also describe how it is sorted (alphabetically by key?)

ddasilva replied:
Good catch on being clear! I meant that the dictionary keys are ordered by the order the corresponding field appears in the packet definition (which is not necessarily alphabetical). Previously an OrderedDict was use to force the order, since older versions of dict didn't make guarantees on keeping insertion order.  Now,  the order of a dict's keys are the order they are inserted in (which is why you don't seen an explicit call to `sort()`).

ehsteve commented:
Reading the code more closely, it does seem like this would work for variable length packets. I was wrong in that the packets are not parsed. Only the ap_id and the packet length.

ddasilva commented:
Thanks for the comments everyone! I will get to addressing these issues. I really appreciate the help and support you've all shown in this merge request.

@ehsteve Since you brought up topic of variable length packets-- this is something I want to support, but I haven't been able to get my hands on any variable length packet data (I have never worked with them myself, and no one thus far has had anything they could send me). If you have any data you can share, we can look into supporting them via a `VariableLength` class.

ddasilva commented:
I addressed the comments in this repush over the existing branch. I wrote a test and it works, but it uses the data that @jmbhughes gave me. Do you give me permission to include that in the repo as test data? I can obfuscate the field names and packet names if that helps. It turns out that the test data we have has a lot of internal consistency problems... if I can't use this I'd have to write synthetic data from scratch.

jmbhughes commented:
It's fine to use the data I shared as test data. It's kind of big though (83 MB). So maybe you should only use the first n packets, say 1,000 packets (I'm not sure how big that would be). That still gives an example of a mixed stream and saves space.
